google/cloud/spanner/value.h

```diff
@@ -0,0 +1,179 @@
+
+#include "google/cloud/optional.h"
+#include "google/cloud/spanner/version.h"
+#include <google/protobuf/struct.pb.h>
+#include <google/spanner/v1/type.pb.h>
+#include <ostream>
+#include <string>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+/**
+ * This class represents a type-safe, nullable Spanner value. It's conceptually
+ * similar to a `std::any` except the only allowed types are those supported by
+ * Spanner, and a "null" value (similar to a `std::any` without a value) still
+ * has an associated type that can be queried with `Value::is<T>()`. The
+ * supported types are shown in the following table along with how they map to
+ * the Spanner types (https://cloud.google.com/spanner/docs/data-types):
+ *
+ *   Spanner Type | C++ Type
+ *   -----------------------
+ *   BOOL         | bool
+ *   INT64        | std::int64_t
+ *   FLOAT64      | double
+ *   STRING       | std::string
+ *
+ * This is a regular C++ value type with support for copy, move, equality, etc,
+ * but there is no default constructor because there is no default type.
+ * Callers may create instances by passing any of the supported values (shown
+ * in the table above) to the constructor. "Null" values are created by passing
+ * a disengaged (i.e., no value) `google::cloud::optional<T>` to the
+ * constructor. Passing a `google::cloud::optional<T>` that contains a value to
+ * the constructor creates a non-null Value and is equivalent to passing the
+ * value without the optional.
+ *
+ * Example with a non-null value:
+ *
+ *   std::string msg = "hello";
+ *   spanner::Value v(msg);
+ *   assert(v.is<std::string>());
+ *   assert(!v.is_null());
+ *   std::string copy = v.get<std::string>();
+ *   assert(msg == copy);
+ *
+ * Example with a null:
+ *
+ *   spanner::Value v(google::cloud::optional<std::int64_t>{});
+ *   assert(v.is<std::int64_t>());
+ *   assert(v.is_null());
+ */
+class Value {
+ public:
+  Value() = delete;
+
+  // Copy and move.
+  Value(Value const&) = default;
+  Value(Value&&) = default;
+  Value& operator=(Value const&) = default;
+  Value& operator=(Value&&) = default;
+
+  // Constructs a non-null instance with the specified value and type.
+  explicit Value(bool v);
+  explicit Value(std::int64_t v);
+  explicit Value(double v);
+  explicit Value(std::string v);
+
+  // Constructs a possibly null instance from the optional value and type.
+  template <typename T>
+  explicit Value(google::cloud::optional<T> opt) {
+    if (opt) {
+      *this = Value(*opt);
+    } else {
+      *this = Value(T{});
+      value_.set_null_value(google::protobuf::NullValue::NULL_VALUE);
+    }
+  }
+
+  friend bool operator==(Value a, Value b);
+  friend bool operator!=(Value a, Value b) { return !(a == b); }
+
+  // Returns true if there is no contained value.
+  bool is_null() const;
+
+  // Returns true if the contained value is of the specified type `T`. All
```

Comment:
I think you want a Doxygen block here:

```C++
/**
 * Returns true if the contained value is of the specified type `T`.
 *
 * All `Value` instances ...
... ...
 */
```

Note how the first sentence is on its own, that makes it the `@brief` description for the function:

http://www.doxygen.nl/manual/commands.html#cmdbrief

Reply:
Done. I think. FWIW, if there's a good doc showing your preferred doxygen style, that'd be great. Or if there's a good tutorial, that'd be helpful too. Otherwise, feel free to be nitpicky until I learn :-)

Reply:
The standard documentation is pretty good.

http://www.doxygen.nl/manual/docblocks.html

I have settled on these things:

1. All types, functions, macros, namespaces in the public API get documentation, even "obvious" classes and functions.
1. Prefer `///` for types or functions that just have a one-liner documentation line.
1. For types or functions that get a long comment start with: (a) a one sentence description of that the thing is or does, then (b) the rest of the documentation. In other words, we use `JAVADOC_AUTOBRIEF=yes`.
1. We use `/** ... */` for long comments, we do not use `/*! ... */`.
1. Use `@` in preference `\` for commands (e.g. `@param` or `@return` instead of `\param` or `\return), no good reason, we just started with those...
1. For functions if you document one parameter you need to document all of them.
1. For functions document add the `@return` section if non-obvious.
1. Large important classes get large documentation, including links to the relevant docs in `cloud.google.com`, but add those late in the development cycle.
1. Using Doxygen for private members is entirely optional, we do not extract this documentation, but it is easier to use the same style for all comments.

Reply:
Yes, this! Carlos's doxygen style tips. Thanks. I made a few updates to the doxygen comments for this file based on this.